streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionSendProcessorSupplier.java

```diff
@@ -115,102 +116,93 @@ public void process(final Record<K, Change<V>> record) {
-                                    currentHash,
-                                    DELETE_KEY_NO_PROPAGATE,
-                                    record.key(),
-                                    partition
-                                )));
-                        //Add to the newKey's state store. Additionally, propagate null if no FK is found there,
-                        //since we must "unset" any output set by the previous FK-join. This is true for both INNER
-                        //and LEFT join.
+                        forward(record, oldForeignKey, DELETE_KEY_NO_PROPAGATE);
                     }
-                    context().forward(
-                        record.withKey(newForeignKey)
-                            .withValue(new SubscriptionWrapper<>(
-                                currentHash,
-                                PROPAGATE_NULL_IF_NO_FK_VAL_AVAILABLE,
-                                record.key(),
-                                partition
-                            )));
+                    //Add to the newKey's state store. Additionally, propagate null if no FK is found there,
+                    //since we must "unset" any output set by the previous FK-join. This is true for both INNER
+                    //and LEFT join.
+                    forward(record, newForeignKey, PROPAGATE_NULL_IF_NO_FK_VAL_AVAILABLE);
                 } else {
-                    //A simple propagatable delete. Delete from the state store and propagate the delete onwards.
-                    context().forward(
-                        record.withKey(oldForeignKey)
-                           .withValue(new SubscriptionWrapper<>(
-                               currentHash,
-                               DELETE_KEY_AND_PROPAGATE,
-                               record.key(),
-                               partition
-                           )));
+                    forward(record, oldForeignKey, DELETE_KEY_AND_PROPAGATE);
                 }
             } else if (record.value().newValue != null) {
-                //change.oldValue is null, which means it was deleted at least once before, or it is brand new.
-                //In either case, we only need to propagate if the FK_VAL is available, as the null from the delete would
-                //have been propagated otherwise.
-
-                final SubscriptionWrapper.Instruction instruction;
-                if (leftJoin) {
-                    //Want to send info even if RHS is null.
-                    instruction = PROPAGATE_NULL_IF_NO_FK_VAL_AVAILABLE;
-                } else {
-                    instruction = PROPAGATE_ONLY_IF_FK_VAL_AVAILABLE;
-                }
-                final KO newForeignKey = foreignKeyExtractor.apply(record.value().newValue);
+                final KO newForeignKey = record.value().newValue == null ? null : foreignKeyExtractor.apply(record.value().newValue);
```

Comment:
this seems redundant with line 165. That check the `newValue` is not `null`

Reply:
indeed!, thanks, removed.